devices: make device name field optional

gmarull opened the pull request:
With the adoption of DEVICE_DT_GET() device names are in many cases
optional and used only for stuff like debugging or in shells. This patch
makes struct device name field optional so that ROM can be saved in case
user can afford it. This means that the name field in struct device
should no longer be accessed directly but using a new function:
device_name_get().

Note that the option controlling such feature is
CONFIG_DEVICE_STORE_NAME, enabled by default.

Some quick numbers using `hello_world`:

`nrf52840dk_nrf52840`: $\Delta_{\textrm{FLASH}} = -100~\textrm{bytes}$
`nucleo_h743zi`: $\Delta_{\textrm{FLASH}} = -272~\textrm{bytes}$

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
devices: make device name field optional

With the adoption of DEVICE_DT_GET() device names are in many cases
optional and used only for stuff like debugging or in shells. This patch
makes struct device name field optional so that ROM can be saved in case
user can afford it. This means that the name field in struct device
should no longer be accessed directly but using a new function:
device_name_get().

Note that the option controlling such feature is
CONFIG_DEVICE_STORE_NAME, enabled by default.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
devices: use device_name_get

Since struct device name field is now optional, use device_name_get.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: shells: s/device_name_get/cmd_device_name_get

Rename the function to avoid clashed with the upcoming device_name_get.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: sensor: fdc2x1x: fix device NULL check

sd_gpio was checked for NULL using its name, so the condition was always
true. Check if the port (device) itself is NULL instead.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

erwango commented:
Nice. Can you update `doc/develop/optimizations/footprint.rst` to make this option somehow visible to users looking for footprint improvements ?

galak commented on kernel/Kconfig:
```diff
@@ -951,6 +951,14 @@ config HAS_DYNAMIC_DEVICE_HANDLES
 	  Hidden option that makes possible to manipulate device handles at
 	  runtime.
 
+config DEVICE_STORE_NAME
```
We need to figure out what this means for the SHELL if CONFIG_DEVICE_STORE_NAME=n

galak replied:
Should `device_get_binding()` even exist if `CONFIG_DEVICE_STORE_NAME=n`, should `device_get_name`?

gmarull replied:
I think they need to exist so that we don't require code changes to enable/disable it. device_get_binding() could just return NULL in that case, though.

galak replied:
why, if you are setting this option, you clearly are concerned about space, so why have the added space taken up in the system?  Seems like we should `ifdef` out the name field in `struct device`.

gmarull replied:
this PR already ifdeffes out `struct device` name field. We can simplify `device_get_binding` to return NULL if preferred.

mbolivar-nordic replied:
Seems my above concern is already discussed here; I agree with @galak on this one, at least for the default.

galak reviewed (changes requested):
Need to figure out for various places that still utilize "name" what it means if `CONFIG_DEVICE_STORE_NAME=n`.  Are those features disabled?  (SHELL, MODBUS, etc)?

gmarull commented:
> Need to figure out for various places that still utilize "name" what it means if `CONFIG_DEVICE_STORE_NAME=n`. Are those features disabled? (SHELL, MODBUS, etc)?

As mentioned, this is an option that not all users will be able to use. But it is there in case you can (it can be useful e.g. for production builds).

galak commented:
> > 
> 
> Afaik, I've replaced all direct usages with device_name_get(). Another option would be to not store anything to `dev->name`, but you're still wasting a pointer per device that does nothing. Btw, not all shells require device_get_binding (e.g. adc), even though it is common, I don't see any device_get_binding under modbus subsys.

`modbus_iface_get_by_name` uses name to match, so it will no longer function correctly.

gmarull commented:
> > > 
> > 
> > 
> > Afaik, I've replaced all direct usages with device_name_get(). Another option would be to not store anything to `dev->name`, but you're still wasting a pointer per device that does nothing. Btw, not all shells require device_get_binding (e.g. adc), even though it is common, I don't see any device_get_binding under modbus subsys.
> 
> `modbus_iface_get_by_name` uses name to match, so it will no longer function correctly.

I think that relies on an independently stored `iface_name`, not on `struct device` name, @jfischer-no

galak commented:
> I think that relies on an independently stored `iface_name`, not on `struct device` name, @jfischer-no

```
subsys/modbus/modbus_core.c:            .iface_name = DEVICE_DT_NAME(DT_DRV_INST(inst)),\
subsys/modbus/modbus_core.c:            .iface_name = "RAW_"#x,                         \
```

gmarull commented:
> > I think that relies on an independently stored `iface_name`, not on `struct device` name, @jfischer-no
> 
> ```
> subsys/modbus/modbus_core.c:            .iface_name = DEVICE_DT_NAME(DT_DRV_INST(inst)),\
> subsys/modbus/modbus_core.c:            .iface_name = "RAW_"#x,                         \
> ```

and that ends up expanding to the actual string in quotes, eg. `"modbus0"`, so not related to `struct device` name except that they take the same ref.

gmarull committed:
doc: footprint: inform about CONFIG_DEVICE_STORE_NAME

CONFIG_DEVICE_STORE_NAME has the potential to reduce ROM usage in case
application images do not make use of device_get_binding(), so mention
it in the footprint documentation.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

MaureenHelm commented:
Sensor drivers and samples look good.

We need a test that builds with `CONFIG_DEVICE_STORE_NAME=n`

gmarull commented:
> Sensor drivers and samples look good.
> 
> We need a test that builds with `CONFIG_DEVICE_STORE_NAME=n`

any suggestion? not sure where to test this

anangl commented on samples/boards/reel_board/mesh_badge/src/reel_board.c:
```diff
@@ -185,15 +185,15 @@ static int add_hello(uint16_t addr, const char *name)
 			stat->addr = addr;
-			strncpy(stat->name, name, sizeof(stat->name) - 1);
+			strncpy(stat->name, name, sizeof(stat) - 1);
 			stat->hello_count = 1U;
 			stat_count++;
 			return i;
 		}
 
 		if (stat->addr == addr) {
 			/* Update name, incase it has changed */
-			strncpy(stat->name, name, sizeof(stat->name) - 1);
+			strncpy(stat->name, name, sizeof(stat) - 1);
```
These changes seem to be undesirable.

gmarull replied:
good catch, thanks

galak reviewed (commented):
I'd suggest maybe splitting this PR as the `device_name_get` cleanup could get merged now, and than we can hash out the testcase and details on `CONFIG_DEVICE_STORE_NAME`

MaureenHelm commented:
> > Sensor drivers and samples look good.
> > We need a test that builds with `CONFIG_DEVICE_STORE_NAME=n`
> 
> any suggestion? not sure where to test this

Maybe start with `samples/basic/minimal`? It's not exactly thorough, but it would be better than nothing.

gmarull commented:
> I'd suggest maybe splitting this PR as the `device_name_get` cleanup could get merged now, and than we can hash out the testcase and details on `CONFIG_DEVICE_STORE_NAME`

you mean adding the API without the optional name field changes?

galak commented:
> you mean adding the API without the optional name field changes?

yes.  Since the 95% of this PR is the API and changes to use it.  I think its reasonable to encapsulate the access to `name`.

mbolivar-nordic commented on include/zephyr/device.h:
```diff
@@ -751,10 +777,20 @@ int device_supported_foreach(const struct device *dev,
+{
+	ARG_UNUSED(name);
+
+	return NULL;
```
This feels like a [footgun](https://en.wiktionary.org/wiki/footgun) implementation.

I believe we should not provide a device_get_binding implementation at all, at least by default, if CONFIG_DEVICE_STORE_NAME=n.

But I get why you would want this, so perhaps it should be configurable? Something like...

```
choice
	prompt "device_get_binding() behavior"
	depends on !DEVICE_STORE_NAME

config NO_DEVICE_GET_BINDING
	bool "No implementation"
	help
	  No device_get_binding() implementation will be provided.
	  Use of this function will cause build errors.

config NULL_DEVICE_GET_BINDING
	bool "NULL implementation"
	help
	  A stub device_get_binding() implementation will be provided
	  that returns NULL.

endchoice
```

with NO_DEVICE_GET_BINDING being the default behavior

gmarull replied:
We discussed this yesterday in dev-review, agreement was to disable device_get_binding completely, so you'll get a build failure if you can't really use this option.

gmarull committed:
samples: basic: minimal: disable CONFIG_DEVICE_STORE_NAME

Minimal sample is meant to show the minimum Zephyr ROM footprint.
Disable CONFIG_DEVICE_STORE_NAME since the option targets ROM footprint
reduction as well.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull commented on drivers/flash/flash_shell.c:
```diff
@@ -255,11 +255,11 @@ static int cmd_test(const struct shell *shell, size_t argc, char *argv[])
+SHELL_DYNAMIC_CMD_CREATE(dsub_device_name, dsub_device_name_get);
 
-static void device_name_get(size_t idx, struct shell_static_entry *entry)
+static void dsub_device_name_get(size_t idx, struct shell_static_entry *entry)
```
renamed to `cmd`

gmarull commented on include/zephyr/device.h:
```diff
@@ -459,6 +467,22 @@ struct device {
+#ifdef CONFIG_DEVICE_STORE_NAME
+	return dev->name;
+#else
+	return "";
```
adjusted to return `"<device>"`

gmarull commented:
> > > Sensor drivers and samples look good.
> > > We need a test that builds with `CONFIG_DEVICE_STORE_NAME=n`
> > 
> > 
> > any suggestion? not sure where to test this
> 
> Maybe start with `samples/basic/minimal`? It's not exactly thorough, but it would be better than nothing.

disabled the option in this sample

mbolivar-nordic commented on doc/develop/optimizations/footprint.rst:
```diff
@@ -49,6 +49,14 @@ the running application and to provide means for debugging and error handling:
+
+:kconfig:option:`CONFIG_DEVICE_STORE_NAME`
+  When this option is disabled, devices will not have a unique human readable
+  name (:c:member:`device.name`). When obtaining device references at runtime
+  (i.e. :c:func:`device_get_binding`), this option can't be used as name is
+  used to find devices.
```
I think something like this would be a bit clearer:

```
If you set this option to ``n``, you will no longer be able to use
:c:func:`device_get_binding`, because it looks up devices by name.
Use :c:macro:`DEVICE_DT_GET` or another related macro instead.
```

gmarull replied:
done, thanks

mbolivar-nordic reviewed (commented):
I'm all for the dev-review solution. One suggested doc improvement, but this looks good for v3.3 to me.

nvlsianpu commented on include/zephyr/device.h:
```diff
@@ -443,8 +443,16 @@ struct pm_device;
  */
 struct device {
-	/** Name of the device instance */
+#if defined(CONFIG_DEVICE_STORE_NAME) || defined(__DOXYGEN__)
```
why `CONFIG_DEVICE_STORE_NAME` can't be set for documentation build instead?

gmarull replied:
I think this pattern is better than keeping an endless/hard to maintain list in Doxyfile, it's now used in a few other places.

nvlsianpu replied:
OK. That's good excuse.

nvlsianpu commented on include/zephyr/device.h:
```diff
@@ -756,14 +783,19 @@ int device_supported_foreach(const struct device *dev,
  * the device is not found or if the device with that name's
  * initialization function failed.
  */
+#ifndef CONFIG_DEVICE_STORE_NAME
```
#ifdef ?

gmarull replied:
good catch, fixed
